DB/src/main/java/io/deephaven/db/v2/by/StreamFirstByChunkedOperator.java

```diff
@@ -0,0 +1,237 @@
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.db.tables.Table;
+import io.deephaven.db.tables.select.MatchPair;
+import io.deephaven.db.v2.QueryTable;
+import io.deephaven.db.v2.ShiftAwareListener;
+import io.deephaven.db.v2.sources.*;
+import io.deephaven.db.v2.sources.chunk.Attributes.*;
+import io.deephaven.db.v2.sources.chunk.*;
+import io.deephaven.db.v2.utils.Index;
+import io.deephaven.db.v2.utils.OrderedKeys;
+import io.deephaven.db.v2.utils.ReadOnlyIndex;
+import io.deephaven.db.v2.utils.ShiftedOrderedKeys;
+import io.deephaven.util.SafeCloseableList;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.ref.SoftReference;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A firstBy aggregation operator for stream tables.
+ *
+ * @see Table#STREAM_TABLE_ATTRIBUTE
+ */
+public class StreamFirstByChunkedOperator implements IterativeChunkedAggregationOperator {
+
+    private static final int COPY_CHUNK_SIZE = ArrayBackedColumnSource.BLOCK_SIZE;
+
+    private final int numResultColumns;
+    private final ColumnSource<?>[] inputColumns;
+    private final WritableSource<?>[] outputColumns;
+
+    private final Map<String, ArrayBackedColumnSource<?>> resultColumns;
+
+    private long previousTableSize;
+    private SoftReference<LongArraySource> cachedRedirections;
+    private LongArraySource redirections;
+
+    StreamFirstByChunkedOperator(@NotNull final MatchPair[] resultPairs, @NotNull final Table streamTable) {
+        numResultColumns = resultPairs.length;
+        inputColumns = new ColumnSource[numResultColumns];
+        outputColumns = new WritableSource[numResultColumns];
+        final Map<String, ArrayBackedColumnSource<?>> resultColumnsMutable = new LinkedHashMap<>(numResultColumns);
+        for (int ci = 0; ci < numResultColumns; ++ci) {
+            final MatchPair resultPair = resultPairs[ci];
+            final ColumnSource<?> streamSource = streamTable.getColumnSource(resultPair.left());
+            final ArrayBackedColumnSource<?> resultSource = ArrayBackedColumnSource.getMemoryColumnSource(0, streamSource.getType(), streamSource.getComponentType());
+            resultColumnsMutable.put(resultPair.left(), resultSource);
+            inputColumns[ci] = ReinterpretUtilities.maybeConvertToPrimitive(streamSource);
+            outputColumns[ci] = (WritableSource<?>) ReinterpretUtilities.maybeConvertToPrimitive(resultSource);
+            Assert.eq(inputColumns[ci].getChunkType(), "inputColumns[ci].getChunkType()", outputColumns[ci].getChunkType(), "outputColumns[ci].getChunkType()");
+        }
+        resultColumns = Collections.unmodifiableMap(resultColumnsMutable);
+        cachedRedirections = new SoftReference<>(redirections = new LongArraySource());
+    }
+
+    @Override
+    public void ensureCapacity(final long tableSize) {
+        redirections.ensureCapacity(tableSize - previousTableSize);
```

Comment:
This is subtle enough that you should have a comment.  Maybe a comment on the overall strategy of knowing that you can only have new things when they are bigger than previousTableSize.

Reply:
Sufficiently clear?
```
    /**
     * <p>The next destination slot that we expect to be used.
     * <p>Any destination less than this one can safely be ignored while processing adds since the first row can never
     * change once a destination has been created given that we ignore removes.
     */
    private long nextDestination;

    /**
     * <p>The first destination that we used on the current step (if we used any). At the very beginning of a step, this
     * is equivalent to {@link #nextDestination} and also the result table's size.
     * <p>We use this as an offset shift for {@code redirections}, so that {@code redirections} only needs to hold first
     * source keys for newly-added destinations, rather than the entire space.
     * <p>At the end of a step, this is updated to prepare for the next step.
     */
    private long firstDestinationThisStep;
```